src/System.Text.Json/System/Text/Json/JsonReader.cs

```diff
@@ -1,383 +1,836 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
-using System.Text.Utf8;
+using System.Runtime.CompilerServices;
 
 namespace System.Text.Json
 {
-    public struct JsonReader : IDisposable
+    public struct JsonReader
     {
-        private Utf8String _str;
-        private int _index;
-        private int _insideObject;
-        private int _insideArray;
-        public JsonTokenType TokenType;
-        private bool _jsonStartIsObject;
-
-        public enum JsonTokenType
+        // We are using a ulong to represent our nested state, so we can only go 64 levels deep.
+        private const int MaxDepth = sizeof(ulong) * 8;
```

Comment:
What is our plan to work around this limit or is this a constraint we want to stick with?